source/client/process_impl.cc

```diff
@@ -338,6 +338,71 @@ getUserDefinedFactoryConfigPairs(const Options& options) {
+ * being mapped by Worker to being mapped by Plugin.
+ *
+ * @param user_defined_outputs_by_plugin The map that this function will collect the worker data
+ * into, maps plugin name to the set of results for that plugin.
+ * @param worker_user_defined_outputs The per worker results to collect and organize.
+ */
+void collectUserDefinedOutputs(
+    absl::flat_hash_map<std::string, std::vector<nighthawk::client::UserDefinedOutput>>*
+        user_defined_outputs_by_plugin,
+    const std::vector<nighthawk::client::UserDefinedOutput>& worker_user_defined_outputs) {
+  for (const nighthawk::client::UserDefinedOutput& user_defined_output :
+       worker_user_defined_outputs) {
+    std::vector<nighthawk::client::UserDefinedOutput> cross_worker_results_for_plugin =
+        (user_defined_outputs_by_plugin->contains(user_defined_output.plugin_name())
+             ? user_defined_outputs_by_plugin->find(user_defined_output.plugin_name())->second
+             : std::vector<nighthawk::client::UserDefinedOutput>{});
+    cross_worker_results_for_plugin.emplace_back(user_defined_output);
+    user_defined_outputs_by_plugin->insert_or_assign(user_defined_output.plugin_name(),
+                                                     cross_worker_results_for_plugin);
+  }
+}
+
+/**
+ * For each provided user defined output plugin factory, aggregates all of its corresponding results
+ * into a global user defined output.
+ *
+ * @param user_defined_outputs_by_plugin A map of plugin name to the set of collected results for
+ * that plugin.
+ * @param user_defined_output_factories A vector of the plugin factories used in this execution
+ * process.
+ * @return std::vector<nighthawk::client::UserDefinedOutput> The aggregated global results for each
+ * plugin.
+ */
+std::vector<nighthawk::client::UserDefinedOutput> compileGlobalUserDefinedPluginOutputs(
+    const absl::flat_hash_map<std::string, std::vector<nighthawk::client::UserDefinedOutput>>&
+        user_defined_outputs_by_plugin,
+    const std::vector<UserDefinedOutputConfigFactoryPair>& user_defined_output_factories) {
+  std::vector<nighthawk::client::UserDefinedOutput> global_outputs;
+  for (const UserDefinedOutputConfigFactoryPair& config_factory_pair :
+       user_defined_output_factories) {
+    UserDefinedOutputPluginFactory* factory = config_factory_pair.second;
+    nighthawk::client::UserDefinedOutput global_output;
+    global_output.set_plugin_name(factory->name());
+
+    auto it = user_defined_outputs_by_plugin.find(factory->name());
+    if (it != user_defined_outputs_by_plugin.end()) {
+      absl::StatusOr<Envoy::ProtobufWkt::Any> global_output_any =
+          factory->AggregateGlobalOutput(it->second);
+      if (global_output_any.ok()) {
+        *global_output.mutable_typed_output() = *global_output_any;
+      } else {
+        *global_output.mutable_error_message() = global_output_any.status().ToString();
+      }
+    } else {
+      throw NighthawkException(
```

Comment:
Why are we choosing to throw an exception and terminate Nighthawk, instead of storing this as an error_message and allowing it to continue?

If we choose to continue, the user will get the benefit of receiving the measurements collected from the execution.